aclk/aclk_lws_wss_client.c

```diff
@@ -0,0 +1,340 @@
+{
+	aclk_lws_mutex_lock(&inst->read_buf_mutex);
+	_aclk_lws_wss_read_buffer_clear(inst->read_ringbuffer);
+	aclk_lws_mutex_unlock(&inst->read_buf_mutex);
+	aclk_lws_mutex_lock(&inst->write_buf_mutex);
+	_aclk_lws_wss_write_buffer_clear(&inst->write_buffer_head);
+	aclk_lws_mutex_unlock(&inst->write_buf_mutex);
+}
+
+static const struct lws_protocols protocols[] = {
+	{
+		"aclk-wss",
+		aclk_lws_wss_callback,
+		sizeof(struct aclk_lws_wss_perconnect_data),
+		0, 0, 0, 0
+	},
+	{ NULL, NULL, 0, 0, 0, 0, 0 }
+};
+
+static void aclk_lws_wss_log_divert(int level, const char *line) {
+	switch(level){
+		case LLL_ERR:
+			error("Libwebsockets Error: %s", line);
+			break;
+		case LLL_WARN:
+			debug(D_ACLK, "Libwebsockets Warn: %s", line);
+			break;
+		default:
+			error("Libwebsockets try to log with unknown log level (%d), msg: %s", level, line);
+	}
+}
+
+struct aclk_lws_wss_engine_instance* aclk_lws_wss_client_init (const struct aclk_lws_wss_engine_callbacks *callbacks, const char *target_hostname, int target_port) {
+	static int lws_logging_initialized = 0;
+	struct lws_context_creation_info info;
+	struct aclk_lws_wss_engine_instance *inst;
+
+	if(unlikely(!lws_logging_initialized)) {
+		lws_set_log_level(LLL_ERR | LLL_WARN, aclk_lws_wss_log_divert);
+		lws_logging_initialized = 1;
+	}
+
+	if(!callbacks || !target_hostname)
+		return NULL;
+
+	inst = callocz(1, sizeof(struct aclk_lws_wss_engine_instance));
+
+	inst->host = target_hostname;
+	inst->port = target_port;
+
+	memset(&info, 0, sizeof(struct lws_context_creation_info));
+    info.options = LWS_SERVER_OPTION_DO_SSL_GLOBAL_INIT;
+	info.port = CONTEXT_PORT_NO_LISTEN;
+    info.protocols = protocols;
+	info.user = inst;
+	
+	inst->lws_context = lws_create_context(&info);
+	inst->callbacks = *callbacks;
+
+	aclk_lws_mutex_init(&inst->write_buf_mutex);
+	aclk_lws_mutex_init(&inst->read_buf_mutex);
+
+	inst->read_ringbuffer = lws_ring_create(1, ACLK_LWS_WSS_RECV_BUFF_SIZE_BYTES, NULL);
+	if(!inst->read_ringbuffer)
+		goto failure_cleanup;
+
+	return inst;
+
+failure_cleanup:
+	lws_context_destroy(inst->lws_context);
+	free(inst);
+	return NULL;
+}
+
+void aclk_lws_wss_client_destroy(struct aclk_lws_wss_engine_instance* inst) {
+	lws_context_destroy(inst->lws_context);
+	inst->lws_context = NULL;
+	inst->lws_wsi = NULL;
+
+	aclk_lws_wss_clear_io_buffers(inst);
+
+#ifdef ACLK_LWS_MOSQUITTO_IO_CALLS_MULTITHREADED
+	pthread_mutex_destroy(&inst->write_buf_mutex);
+	pthread_mutex_destroy(&inst->read_buf_mutex);
+#endif
+}
+
+void _aclk_wss_connect(struct aclk_lws_wss_engine_instance *inst){
+	struct lws_client_connect_info i;
+
+	memset(&i, 0, sizeof(i));
+	i.context = inst->lws_context;
+	i.port = inst->port;
+	i.address = inst->host;
+	i.path = "/mqtt";
+	i.host = inst->host;
+	i.protocol = "mqtt";
+	if(aclk_lws_wss_use_ssl)
+//TODO!!!!!! REMOVE ME 
+#define ACLK_SSL_ALLOW_SELF_SIGNED 1
+#ifdef ACLK_SSL_ALLOW_SELF_SIGNED
```

Comment:
We could replace this with `#ifdef MAINTAINER_MODE` and it will still be useful for internal testing. I notice that maintainer mode doesn't actually do anything in the `configure.ac` though so you would need to add a def inside the test block.

Reply:
i removed it in PR #7988 as that is where it belongs